adage/analyze/api.py

```diff
@@ -212,6 +213,36 @@ class Meta:
         queryset = Node.objects.all()
         resource_name = 'node'
         allowed_methods = ['get']
+        filtering = {
+            'name': ('exact', 'in', ),
+            'heavy_genes': ('exact', ),  # New filter, see apply_filters().
+        }
+
+    def apply_filters(self, request, applicable_filters):
+        object_list = super(NodeResource, self).apply_filters(
+            request, applicable_filters)
+        heavy_genes = request.GET.get('heavy_genes', None)
+        if heavy_genes:
+            # Instead of relying on tastypie/resources.py to catch the
+            # ValueError exception that may be raised in this function,
+            # we catch the one that may be raised by int() below and
+            # raise a customized BadRequest exception with more details.
+            try:
+                # Convert genes to a set of integers so that the
+                # duplicate(s) will be removed implicitly.
+                query_genes = {int(id) for id in heavy_genes.split(',')}
+            except ValueError:
+                raise BadRequest("Invalid gene IDs: %s" % heavy_genes)
+
+            for (i, q) in enumerate(query_genes):
```

Comment:
Instead of this loop, can this be `Participation.objects.filter(gene_id__in=query_genes).values_list('node_id', flat=True)`? You could make it a set if you wanted to.